[android-navigation] Bottom drawer definition crash on resume after system initiated process death

# AniTrend Issue Guidelines

Before opening a new issue, please take a moment to review our [**community guidelines**](https://github.com/AniTrend/anitrend-v2/blob/master/CONTRIBUTING.md) to make the contribution process easy and effective for everyone involved.

**You may find an answer in already closed issues**:
https://github.com/AniTrend/anitrend-v2/issues?q=is%3Aissue+is%3Aclosed


## Description of Bug
<!-- A clear and concise description of what the bug is. e.g. -->

Leaving the application in background for a long time, in my case a few hours occasionally triggers a crash. This might vary from device to device based on available system memory and vendor power saving configuration. It seems like a view model creation issue happens post resuming the application from process death, hinting towards a failure to recreate `saveInstanceBundle`

## Reproduction Steps
<!-- In detail, please explain how and what can be done to reproduce this bug. -->
- Open anitrend and make sure you're on the main screen as seen in the screenshot below
- [Simulate process death](https://zhuinden.medium.com/you-can-actually-simulate-process-death-by-putting-your-application-in-the-background-with-home-89fc768a96c0)

> Simulating process death was able to reproduce the crash sometimes

<img src="https://images.zenhubusercontent.com/5f6313bf4c8b9e21a739d6b3/c0fc3ae6-60c1-4417-a8d4-da4be4436387" width="250px" />

## Screenshots/Videos
<!-- Screenshots or videos usally paint a better picutre, if you have non please remove this heading -->


## Device & Build Information
<!-- Please provide any relevant information about your device. This is important in case the issue is not reproducible except for under certain conditions. -->

Not device dependent, affects all the current alpha versions

* OS Version: 10
* App Version: 2.0.0-alpha40
* Phone Model: POCO F1
* Phone Brand: Xiaomi

Donwload Source:
- [ ] Playstore
- [ ] Github
- [x] Other (built from source)

## Additional Context
<!--
Providing context helps us come up with a solution that is most useful in the real world, 
also include any logs if you have any in this section, if not please remove this section
-->

Further investigations are still required to assure that we can reproduce this reliably 

```sh
2022-07-23 09:42:15.993 14957-14957/co.anitrend E/ExceptionCrashHandler: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
2022-07-23 09:42:15.993 14957-14957/co.anitrend E/ExceptionCrashHandler:     at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:549)
2022-07-23 09:42:15.993 14957-14957/co.anitrend E/ExceptionCrashHandler:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
2022-07-23 09:42:15.993 14957-14957/co.anitrend E/ExceptionCrashHandler: Caused by: java.lang.reflect.InvocationTargetException
2022-07-23 09:42:15.993 14957-14957/co.anitrend E/ExceptionCrashHandler:     at java.lang.reflect.Method.invoke(Native Method)
2022-07-23 09:42:15.993 14957-14957/co.anitrend E/ExceptionCrashHandler:     at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:539)
2022-07-23 09:42:15.993 14957-14957/co.anitrend E/ExceptionCrashHandler: 	... 1 more
2022-07-23 09:42:15.993 14957-14957/co.anitrend E/ExceptionCrashHandler: Caused by: org.koin.core.error.InstanceCreationException: Could not create instance for [Factory:'co.anitrend.navigation.drawer.component.viewmodel.BottomDrawerViewModel']
2022-07-23 09:42:15.993 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.core.instance.InstanceFactory.create(InstanceFactory.kt:60)
2022-07-23 09:42:15.993 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.core.instance.FactoryInstanceFactory.get(FactoryInstanceFactory.kt:38)
2022-07-23 09:42:15.993 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.core.registry.InstanceRegistry.resolveInstance$koin_core(InstanceRegistry.kt:110)
2022-07-23 09:42:15.993 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.core.scope.Scope.resolveValue(Scope.kt:254)
2022-07-23 09:42:15.993 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.core.scope.Scope.resolveInstance(Scope.kt:241)
2022-07-23 09:42:15.993 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.core.scope.Scope.get(Scope.kt:204)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.core.scope.Scope.getOrNull(Scope.kt:172)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.core.scope.Scope.findInOtherScope(Scope.kt:289)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.core.scope.Scope.resolveValue(Scope.kt:269)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.core.scope.Scope.resolveInstance(Scope.kt:241)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.core.scope.Scope.get(Scope.kt:204)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at androidx.lifecycle.StateViewModelFactory.create(StateViewModelFactory.kt:17)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at androidx.lifecycle.AbstractSavedStateViewModelFactory.create(AbstractSavedStateViewModelFactory.java:69)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:185)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:150)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.androidx.viewmodel.ViewModelResolverKt.resolveInstance(ViewModelResolver.kt:15)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.androidx.viewmodel.scope.ScopeExtKt.getViewModel(ScopeExt.kt:67)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.androidx.viewmodel.scope.ScopeExtKt.getViewModel(ScopeExt.kt:53)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.androidx.viewmodel.ext.android.SavedStateRegistryOwnerExtKt.getStateViewModel(SavedStateRegistryOwnerExt.kt:73)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at co.anitrend.navigation.drawer.component.content.BottomDrawerContent$special$$inlined$stateViewModel$default$1.invoke(SavedStateRegistryOwnerExt.kt:75)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at co.anitrend.navigation.drawer.component.content.BottomDrawerContent$special$$inlined$stateViewModel$default$1.invoke(SavedStateRegistryOwnerExt.kt:42)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at kotlin.UnsafeLazyImpl.getValue(Lazy.kt:81)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at co.anitrend.navigation.drawer.component.content.BottomDrawerContent.getViewModel(BottomDrawerContent.kt:71)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at co.anitrend.navigation.drawer.component.content.BottomDrawerContent.setUpViewModelObserver(BottomDrawerContent.kt:247)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at co.anitrend.arch.ui.fragment.SupportFragment.onViewCreated(SupportFragment.kt:125)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at co.anitrend.navigation.drawer.component.content.BottomDrawerContent.onViewCreated(BottomDrawerContent.kt:260)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at androidx.fragment.app.Fragment.performViewCreated(Fragment.java:3019)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:551)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:261)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at androidx.fragment.app.FragmentStore.moveToExpectedState(FragmentStore.java:113)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1374)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:2841)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at androidx.fragment.app.FragmentManager.dispatchActivityCreated(FragmentManager.java:2784)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at androidx.fragment.app.FragmentController.dispatchActivityCreated(FragmentController.java:262)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at androidx.fragment.app.FragmentActivity.onStart(FragmentActivity.java:478)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at androidx.appcompat.app.AppCompatActivity.onStart(AppCompatActivity.java:246)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1433)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at android.app.Activity.performStart(Activity.java:7923)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at android.app.ActivityThread.handleStartActivity(ActivityThread.java:3332)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at android.app.servertransaction.TransactionExecutor.performLifecycleSequence(TransactionExecutor.java:221)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at android.app.servertransaction.TransactionExecutor.cycleToPath(TransactionExecutor.java:201)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:173)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2044)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at android.os.Handler.dispatchMessage(Handler.java:107)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at android.os.Looper.loop(Looper.java:224)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at android.app.ActivityThread.main(ActivityThread.java:7562)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler: 	... 3 more
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler: Caused by: org.koin.core.error.NoBeanDefFoundException: |- No definition found for class:'androidx.lifecycle.SavedStateHandle'. Check your definitions!
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.core.scope.Scope.throwDefinitionNotFound(Scope.kt:304)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.core.scope.Scope.resolveValue(Scope.kt:274)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.core.scope.Scope.resolveInstance(Scope.kt:241)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.core.scope.Scope.get(Scope.kt:204)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at co.anitrend.navigation.drawer.koin.ModulesKt$viewModelModule$1$1.invoke(Modules.kt:113)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at co.anitrend.navigation.drawer.koin.ModulesKt$viewModelModule$1$1.invoke(Modules.kt:52)
2022-07-23 09:42:15.994 14957-14957/co.anitrend E/ExceptionCrashHandler:     at org.koin.core.instance.InstanceFactory.create(InstanceFactory.kt:53)
```

--- a/android-navigation/src/main/kotlin/co/anitrend/navigation/drawer/koin/Modules.kt
+++ b/android-navigation/src/main/kotlin/co/anitrend/navigation/drawer/koin/Modules.kt
@@ -18,6 +18,7 @@
 package co.anitrend.navigation.drawer.koin
 
 import co.anitrend.core.koin.helper.DynamicFeatureModuleHelper
+import co.anitrend.core.koin.scope.AppScope
 import co.anitrend.navigation.NavigationDrawerRouter
 import co.anitrend.navigation.drawer.action.provider.viewmodel.NotificationProviderViewModel
 import co.anitrend.navigation.drawer.action.provider.viewmodel.state.AuthenticatedUserState
@@ -75,18 +76,20 @@ private val viewModelModule = module {
 }
 
 private val fragmentModule = module {
-	fragment {
-		BottomDrawerContent(
-			navigationAdapter = NavigationAdapter(
-				resources = androidContext().resources,
-				stateConfiguration = get()
-			),
-			accountAdapter = AccountAdapter(
-				resources = androidContext().resources,
-				stateConfiguration = get()
-			),
-		)
-	} bind INavigationDrawer::class
+	scope(AppScope.BOTTOM_NAV_DRAWER.qualifier) {
+		fragment {
+			BottomDrawerContent(
+				navigationAdapter = NavigationAdapter(
+					resources = androidContext().resources,
+					stateConfiguration = get()
+				),
+				accountAdapter = AccountAdapter(
+					resources = androidContext().resources,
+					stateConfiguration = get()
+				),
+			)
+		} bind INavigationDrawer::class
+	}
 }
 
 private val featureModule = module {

--- a/app-core/src/main/kotlin/co/anitrend/core/extensions/ScopeExtensions.kt
+++ b/app-core/src/main/kotlin/co/anitrend/core/extensions/ScopeExtensions.kt
@@ -1,0 +1,38 @@
+/*
+ * Copyright (C) 2022  AniTrend
+ *
+ *     This program is free software: you can redistribute it and/or modify
+ *     it under the terms of the GNU General Public License as published by
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ *
+ *     This program is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ *     You should have received a copy of the GNU General Public License
+ *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package co.anitrend.core.extensions
+
+import co.anitrend.core.koin.scope.contract.ICustomScope
+import org.koin.core.scope.Scope
+import org.koin.mp.KoinPlatformTools
+import timber.log.Timber
+
+/**
+ * Helper to get or create a custom scope
+ *
+ * @return [Scope] of custom type
+ */
+operator fun ICustomScope.invoke(): Scope {
+    val koin = KoinPlatformTools.defaultContext().get()
+    val scope = koin.getOrCreateScope(
+        toString(),
+        qualifier
+    )
+    Timber.v("Retrieving or creating custom scope: $scope")
+    return scope
+}

--- a/app-core/src/main/kotlin/co/anitrend/core/koin/scope/AppScope.kt
+++ b/app-core/src/main/kotlin/co/anitrend/core/koin/scope/AppScope.kt
@@ -1,0 +1,26 @@
+/*
+ * Copyright (C) 2022  AniTrend
+ *
+ *     This program is free software: you can redistribute it and/or modify
+ *     it under the terms of the GNU General Public License as published by
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ *
+ *     This program is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ *     You should have received a copy of the GNU General Public License
+ *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package co.anitrend.core.koin.scope
+
+import co.anitrend.core.koin.scope.contract.ICustomScope
+import org.koin.core.qualifier.Qualifier
+import org.koin.core.qualifier.named
+
+enum class AppScope(override val qualifier: Qualifier) : ICustomScope {
+    BOTTOM_NAV_DRAWER(named("bottom_nav_drawer"))
+}

--- a/app-core/src/main/kotlin/co/anitrend/core/koin/scope/contract/ICustomScope.kt
+++ b/app-core/src/main/kotlin/co/anitrend/core/koin/scope/contract/ICustomScope.kt
@@ -1,0 +1,27 @@
+/*
+ * Copyright (C) 2022  AniTrend
+ *
+ *     This program is free software: you can redistribute it and/or modify
+ *     it under the terms of the GNU General Public License as published by
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ *
+ *     This program is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ *     You should have received a copy of the GNU General Public License
+ *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package co.anitrend.core.koin.scope.contract
+
+import org.koin.core.qualifier.Qualifier
+
+/**
+ * Contract for defining custom scopes
+ */
+interface ICustomScope {
+    val qualifier: Qualifier
+}

--- a/app/src/main/kotlin/co/anitrend/component/screen/MainScreen.kt
+++ b/app/src/main/kotlin/co/anitrend/component/screen/MainScreen.kt
@@ -34,7 +34,9 @@ import co.anitrend.component.action.ShowHideFabStateAction
 import co.anitrend.component.presenter.MainPresenter
 import co.anitrend.component.viewmodel.MainScreenViewModel
 import co.anitrend.core.component.screen.AniTrendScreen
+import co.anitrend.core.extensions.invoke
 import co.anitrend.core.extensions.orEmpty
+import co.anitrend.core.koin.scope.AppScope
 import co.anitrend.core.ui.commit
 import co.anitrend.core.ui.fragmentByTagOrNew
 import co.anitrend.core.ui.inject
@@ -152,6 +154,14 @@ class MainScreen : AniTrendScreen<MainScreenBinding>() {
                 drawerFragmentItem.tag()
             )
         }
+    }
+
+    /**
+     * Can be used to configure custom theme styling as desired
+     */
+    override fun configureActivity() {
+        scope.linkTo(AppScope.BOTTOM_NAV_DRAWER())
+        super.configureActivity()
     }
 
     override fun onCreate(savedInstanceState: Bundle?) {

